File: S03/week04/sejun/SimplifyPath.java
// 5ms Beats 52.70% of users with Java
// 43.37MB Beats 46.70% of users with Java
class Solution {
    private static final String CURRENT_DIR = ".";
    private static final String UP_DIR = "..";
    private static final String EMPTY = "";
    private static final String SLASH = "/";
    public String simplifyPath(String path) {
        String[] pathArr = path.split(SLASH);
        
        List<String> result = new ArrayList<>();
        for (String s : pathArr) {
            if (CURRENT_DIR.equals(s)) {
                continue;
            }

            if (UP_DIR.equals(s)) {
                if (!result.isEmpty()) {
                    result.remove(result.size() - 1);
                }
                continue;
            }

            if (EMPTY.equals(s)) {
                continue;
            }

            result.add(s);
        }

        return SLASH + String.join("/", result);
    }
}

File: S03/week04/seokjoon/SimplifyPath.kt
//195ms, 38mb
class Solution {
    fun simplifyPath(path: String): String {
        val pathItemList = path.split("/")
        val stack = Stack<String>()

        for(pathItem in pathItemList){
            if(pathItem == ".."){
                if(stack.isNotEmpty()){
                    stack.pop()
                }
            }else if(pathItem == "." || pathItem.isEmpty()) continue
            else {
                stack.push(pathItem)   
            }
        }
        
        return "/${stack.joinToString("/")}"
    }
}
